pkg/cloud/services/compute/instance.go

```diff
@@ -277,24 +282,98 @@ func getPortName(instanceName string, opts *infrav1.PortOpts, netIndex int) stri
 	return fmt.Sprintf("%s-%d", instanceName, netIndex)
 }
 
-// applyRootVolume sets a root volume if the root volume Size is not 0.
-func applyRootVolume(opts servers.CreateOptsBuilder, rootVolume *infrav1.RootVolume) servers.CreateOptsBuilder {
-	if rootVolume != nil && rootVolume.Size != 0 {
-		block := bootfromvolume.BlockDevice{
-			SourceType:          bootfromvolume.SourceType(rootVolume.SourceType),
-			BootIndex:           0,
-			UUID:                rootVolume.SourceUUID,
-			DeleteOnTermination: true,
-			DestinationType:     bootfromvolume.DestinationVolume,
-			VolumeSize:          rootVolume.Size,
-			DeviceType:          rootVolume.DeviceType,
-		}
-		return bootfromvolume.CreateOptsExt{
-			CreateOptsBuilder: opts,
-			BlockDevice:       []bootfromvolume.BlockDevice{block},
+func rootVolumeName(instanceName string) string {
+	return fmt.Sprintf("%s-root", instanceName)
+}
+
+func hasRootVolume(rootVolume *infrav1.RootVolume) bool {
+	return rootVolume != nil && rootVolume.Size > 0
+}
+
+func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
+	listOpts := volumes.ListOpts{
+		AllTenants: false,
+		Name:       name,
+		TenantID:   s.projectID,
+	}
+	volumeList, err := s.computeService.ListVolumes(listOpts)
+	if err != nil {
+		return nil, fmt.Errorf("error listing volumes: %w", err)
+	}
+	if len(volumeList) > 1 {
+		return nil, fmt.Errorf("expected to find a single volume called %s; found %d", name, len(volumeList))
+	}
+	if len(volumeList) == 0 {
+		return nil, nil
+	}
+	return &volumeList[0], nil
+}
+
+func (s *Service) getOrCreateRootVolume(eventObject runtime.Object, instanceSpec *InstanceSpec) (*volumes.Volume, error) {
+	rootVolume := instanceSpec.RootVolume
+	if hasRootVolume(rootVolume) {
+		return nil, nil
+	}
+
+	// TODO: Remove SourceType and DeviceType from the API because we don't support customising them
```

Comment:
todo: I think we can put those validation into webhook later on?
as if we check (guess we have another option to delegate this check to openstack itself?) we can always do at API level?

Reply:
These are taken directly from the nova `block_device_mapping_v2` field to create: https://docs.openstack.org/api-ref/compute/?expanded=create-server-detail#create-server. This is a problem as this change stops using this API to create volumes. This is because the Nova API is a proxy API, and doesn't expose the full functionality of Cinder. By policy Nova will not generally add new functionality to proxy APIs. They prefer to remove them in favour of the client making explicit calls, which is what we're doing here.

The possible values of `SourceType` are:
* `blank`: a blank disk
* `image`: initialise from a glance image
* `snapshot`: initialise from a volume snapshot
* (`volume`: attach a specific volume, which isn't relevant here)

`blank` doesn't work for a root volume, so the only values which make any sense are `image` and `snapshot`. As far as I'm aware we only support glance images, so any value other than `image` would be a mistake. We might as well remove `SourceType` entirely.

`DeviceType` relates to how the hypervisor exposes the disk to the guest VM. Potential values include:
* `disk`
* `cdrom`
* `floppy`

Unless somebody makes a K8s distro which fits in 1.44MB or which boots from a Live CD, any value other than `disk` is a mistake. We might as well remove it entirely.

Reply:
yes, I knew those proxy and param pass :) thanks for point that out ~

I am only thinking about reject those at API layer as actually those are bad API request
so instead of fail later, we fail early ~

but another thought comes out ,we have many many places might fit into such check
put all those into API layer might make the validation too big ...

Reply:
You mean adding them to the validating web hook in this PR?

Given that we're about to cut v1alpha5 and hopefully soon v1beta1 I was hoping to slip in the breaking API change of just removing them entirely. I'm still not entirely sure how that works with branching, though, and of course I also need to make the case for breaking the API.

However, would I still want to add it for validating <v1alpha5? 🤔 And if we decide against removing the API entirely in this PR, the validating webhook also sounds like the better option.

As for making the validation too big, it's a very cheap test to execute. I haven't really looked at the validation in detail yet, so I don't know how much code complexity it would add.

Reply:
I would be in favor of removing the API fields that serve no function other than as a trap for users. Its not like there is even a use case or a reason they should exist, so why support and validate them in the first place?